ringbuf/reader.go

```diff
@@ -0,0 +1,255 @@
+package ringbuf
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"runtime"
+	"sync"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal"
+	"github.com/cilium/ebpf/internal/unix"
+)
+
+var (
+	errClosed  = errors.New("ringbuf reader was closed")
+	errDiscard = errors.New("sample discarded")
+	errBusy    = errors.New("sample not committed yet")
+)
+
+func addToEpoll(epollfd, fd int) error {
+
+	event := unix.EpollEvent{
+		Events: unix.EPOLLIN,
+		Fd:     int32(fd),
+	}
+
+	if err := unix.EpollCtl(epollfd, unix.EPOLL_CTL_ADD, fd, &event); err != nil {
+		return fmt.Errorf("can't add fd to epoll: %v", err)
+	}
+	return nil
+}
+
+// ringbufHeader from 'struct bpf_ringbuf_hdr' in kernel/bpf/ringbuf.c
+type ringbufHeader struct {
+	Len   uint32
+	PgOff uint32
+}
+
+type Record struct {
+	RawSample []byte
+}
+
+func readRecordFromRing(ring *ringbufEventRing) (Record, error) {
+	return readRecord(ring)
+}
+
+func readRecord(rd io.Reader) (Record, error) {
+	var header ringbufHeader
+	err := binary.Read(rd, internal.NativeEndian, &header)
+	if err == io.EOF {
+		return Record{}, err
+	}
+
+	if err != nil {
+		return Record{}, fmt.Errorf("can't read event header: %v", err)
+	}
+
+	if header.Len&unix.BPF_RINGBUF_BUSY_BIT != 0 {
+		return Record{}, errBusy
+	}
+
+	dataLen := header.Len << 2
+	dataLen >>= 2
+
+	/* read up to 8 byte alignment */
+	data := make([]byte, (dataLen+7)/8*8)
+	if _, err := io.ReadFull(rd, data); err != nil {
+		return Record{}, fmt.Errorf("can't read sample: %v", err)
+	}
+
+	if header.Len&unix.BPF_RINGBUF_DISCARD_BIT == 0 {
+		return Record{RawSample: data[:dataLen]}, nil
+	}
+
+	return Record{}, errDiscard
+
+}
+
+// Reader allows reading bpf_ringbuf_output
+// from user space.
+type Reader struct {
+	// mu protects read/write access to the Reader structure
+	mu sync.Mutex
+
+	array *ebpf.Map
+	mapFd int
+	ring  *ringbufEventRing
+
+	epollFd     int
+	epollEvents []unix.EpollEvent
+	pollTimeout int
+
+	closeFd int
+	// Ensure we only close once
+	closeOnce sync.Once
+}
+
+// ReaderOptions control the behaviour of the user
+// space reader.
+type ReaderOptions struct {
+	pollTimeout int
```

Comment:
After reading the rest of the code, this probably shouldn't be exposed to users. Since it's only used for RB_NO_WAKEUP tests (and I think those don't necessarily belong in the library) we might not want it at all.

Reply:
@lmb my idea was to at least allow users to busy poll the ring buffer without having to wait of epoll notification first. are you suggesting I drop the whole `pollTimeout` setting and always use `-1`?

Reply:
Yes, unless you have a strong motivation why this is needed.

Reply:
The only motivation would be to keep the functionality aligned with libbpf. The kernel docs suggest this feature in BPF ringbuf as one of the differentiating features compared with perf events:

>  There are a bunch of similarities between perf buffer (`BPF_MAP_TYPE_PERF_EVENT_ARRAY`) and new BPF ring buffer semantics:
>
> - variable-length records;
> - if there is no more space left in ring buffer, reservation fails, no blocking;
> - memory-mappable data area for user-space applications for ease of consumption and high performance;
> - epoll notifications for new incoming data;
> - but still the ability to do busy polling for new data to achieve the lowest latency, if necessary.

https://www.kernel.org/doc/html/latest/bpf/ringbuf.html

Reply:
Maybe I'm missing something. In your implementation, busy polling boils down to passing `0` timeout to EpollWait. Since the perf reader also uses EpollWait we could pass 0 there and get busy polling as well? So my argument is: let's not add it to only one if both can support it. Second, let's get the simple version merged first, we can add busy polling support if needed later.

TBH that list of features seems like a bit of marketing to me. All but the first point apply to perf buffers as well, and arguably variable length records are supported by perf buffers but the BPF API isn't there to make use of them.

Reply:
As far as I can see, perf with `0` timeout would still depend on the kernel to make data available via poll events. My implementation always reads the ring without checking for data availability as a poll event. Ringbuf samples submitted with `BPF_RB_NO_WAKEUP` can be used to skip the whole notification... Source: https://nakryiko.com/posts/bpf-ringbuf/#bpf-ringbuf-data-notification-control

> Second, let's get the simple version merged first, we can add busy polling support if needed later.

This works for me and I can drop busy polling related stuff (including the benchmarks/test cases) from this version.